src/query/src/promql/planner.rs

```diff
@@ -1739,10 +1762,34 @@ impl PromPlanner {
                 }
                 ScalarFunc::DataFusionUdf(Arc::new(Round::scalar_udf()))
             }
+            "rad" => ScalarFunc::DataFusionBuiltin(datafusion::functions::math::radians()),
+            "deg" => ScalarFunc::DataFusionBuiltin(datafusion::functions::math::degrees()),
+            "sgn" => ScalarFunc::DataFusionBuiltin(datafusion::functions::math::signum()),
+            "pi" => {
+                // pi functions doesn't accepts any arguments, needs special processing
+                let fn_expr = DfExpr::ScalarFunction(ScalarFunction {
+                    func: datafusion::functions::math::pi(),
+                    args: vec![],
+                });
+                exprs.push(fn_expr);
 
+                ScalarFunc::GeneratedExpr
+            }
             _ => {
-                if let Some(f) = session_state.scalar_functions().get(func.name) {
+                if let Some(f) = query_engine_state
+                    .session_state()
+                    .scalar_functions()
+                    .get(func.name)
+                {
                     ScalarFunc::DataFusionBuiltin(f.clone())
+                } else if let Some(factory) = query_engine_state.scalar_function(func.name) {
+                    let func_state = query_engine_state.function_state();
+                    let query_ctx = self.table_provider.query_ctx();
+
+                    ScalarFunc::DataFusionUdf(Arc::new(factory.provide(FunctionContext {
+                        state: func_state,
+                        query_ctx: query_ctx.clone(),
+                    })))
```

Comment:
Can we register and get all our functions to session state? They should not be different (at least from the implementation side)

Reply:
No, we can't, because some functions depend on the query context, such as timezone settings. They have to be created dynamically via the function factory.